src/StringTools/StringTools.cs

```diff
@@ -0,0 +1,233 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+#if !NET35
+using Microsoft.Extensions.ObjectPool;
+#endif
+
+namespace Microsoft.StringTools
+{
+    /// <summary>
+    /// A callback type to register with the library. The library will invoke this callback in registration order until
+    /// true is returned. If no callback returns true, the library will fall back to its default interning logic.
+    /// </summary>
+    /// <param name="internableString">The string to be interned.</param>
+    /// <param name="result">The interned string. Set to null and return true if the string should not be interned.</param>
+    /// <returns>True if the <paramref name="internableString" /> was handled by the callback.</returns>
+    public delegate bool TryInternStringDelegate(ref InternableString internableString, out string result);
+
+    public static class Strings
+
+    {
+#if !NET35
+        /// <summary>
+        /// IPooledObjectPolicy used by <cref see="s_stringBuilderPool"/>.
+        /// </summary>
+        private class PooledObjectPolicy : IPooledObjectPolicy<SpanBasedStringBuilder>
+        {
+            /// <summary>
+            /// No need to retain excessively long builders forever.
+            /// </summary>
+            private const int MAX_RETAINED_BUILDER_CAPACITY = 1000;
+
+            /// <summary>
+            /// Creates a new SpanBasedStringBuilder with the default capacity.
+            /// </summary>
+            public SpanBasedStringBuilder Create()
+            {
+                return new SpanBasedStringBuilder();
+            }
+
+            /// <summary>
+            /// Returns a builder to the pool unless it's excessively long.
+            /// </summary>
+            public bool Return(SpanBasedStringBuilder stringBuilder)
+            {
+                if (stringBuilder.Capacity <= MAX_RETAINED_BUILDER_CAPACITY)
```

Comment:
Why this check?

Reply:
The idea was to not return the builder to the pool if it was excessively large. The idea is that if the object is unlikely to be needed again, it is better to throw it away than to keep it forever. I have removed this check, thanks!